internal/datastore/postgres/revisions.go

```diff
@@ -0,0 +1,113 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/shopspring/decimal"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+const (
+	errRevision      = "unable to find revision: %w"
+	errCheckRevision = "unable to check revision: %w"
+)
+
+func (pgd *pgDatastore) HeadRevision(ctx context.Context) (datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "HeadRevision")
+	defer span.End()
+
+	revision, err := pgd.loadRevision(ctx)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+
+	return revisionFromTransaction(revision), nil
+}
+
+func (pgd *pgDatastore) OptimizedRevision(ctx context.Context) (datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "OptimizedRevision")
+	defer span.End()
+
+	var revision uint64
+	if err := pgd.dbpool.QueryRow(
+		datastore.SeparateContextWithTracing(ctx), pgd.optimizedRevisionQuery,
+	).Scan(&revision); err != nil {
+		return datastore.NoRevision, fmt.Errorf(errRevision, err)
+	}
+
+	return revisionFromTransaction(revision), nil
+}
+
+func (pgd *pgDatastore) CheckRevision(ctx context.Context, revision datastore.Revision) error {
+	ctx, span := tracer.Start(ctx, "CheckRevision")
+	defer span.End()
+
+	revisionTx := transactionFromRevision(revision)
+	offset := uint64(pgd.gcWindowInverted.Seconds())
+
+	queryValid := fmt.Sprintf(
```

Comment:
Same question re: parameterization

Reply:
I can fill in most of them at initialization time, and then pass one as an arg.